File: server/webserver/webserver.go
package webserver

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
	"vpn-sandbox/core"
	"vpn-sandbox/utils"

	"github.com/gorilla/mux"
)

type IpInfo map[string]interface{}

var (
	staticDir     = "./static"
	ipInfo        = &ipInfoCache{}
	statusUpdates = newStatusNotifier()
)

const statusHeartbeatInterval = 15 * time.Second

type ipInfoStatus struct {
	Output     IpInfo     `json:"output"`
	ExecutedAt *time.Time `json:"executedAt,omitempty"`
	Event      string     `json:"event"`
	EventAt    time.Time  `json:"eventAt"`
	Stale      bool       `json:"stale"`
}

type ipInfoCache struct {
	mu          sync.RWMutex
	refreshMu   sync.Mutex
	lookup      func(IpInfo) error
	output      IpInfo
	executedAt  time.Time
	lastEvent   string
	lastEventAt time.Time
}

type statusEventListener struct{}

type statusNotifier struct {
	mu          sync.RWMutex
	subscribers map[chan string]struct{}
}

func newStatusNotifier() *statusNotifier {
	return &statusNotifier{subscribers: make(map[chan string]struct{})}
}

func (n *statusNotifier) subscribe() (<-chan string, func()) {
	updates := make(chan string, 1)
	n.mu.Lock()
	n.subscribers[updates] = struct{}{}
	n.mu.Unlock()

	return updates, func() {
		n.mu.Lock()
		delete(n.subscribers, updates)
		n.mu.Unlock()
	}
}

func (n *statusNotifier) notify(event string) {
	n.mu.RLock()
	defer n.mu.RUnlock()
	for subscriber := range n.subscribers {
		select {
		case subscriber <- event:
		default:
		}
	}
}

func (i *ipInfoCache) markEvent(name string, at time.Time) {
	i.mu.Lock()
	i.lastEvent = name
	i.lastEventAt = at
	i.mu.Unlock()
}

func (i *ipInfoCache) store(output IpInfo, executedAt time.Time) {
	i.mu.Lock()
	i.output = output
	i.executedAt = executedAt
	i.mu.Unlock()
}

func (i *ipInfoCache) snapshot() ipInfoStatus {
	i.mu.RLock()
	defer i.mu.RUnlock()

	output := make(IpInfo, len(i.output))
	for key, value := range i.output {
		output[key] = value
	}

	var executedAt *time.Time
	if !i.executedAt.IsZero() {
		value := i.executedAt
		executedAt = &value
	}

	return ipInfoStatus{
		Output:     output,
		ExecutedAt: executedAt,
		Event:      i.lastEvent,
		EventAt:    i.lastEventAt,
		Stale:      i.executedAt.IsZero() || i.executedAt.Before(i.lastEventAt),
	}
}

func (i *ipInfoCache) refresh() bool {
	i.refreshMu.Lock()
	defer i.refreshMu.Unlock()

	lookup := i.lookup
	if lookup == nil {
		lookup = func(output IpInfo) error {
			return utils.GetIpInfo(output)
		}
	}

	updated := IpInfo{}
	if err := lookup(updated); err != nil {
		return false
	}

	i.store(updated, time.Now())
	return true
}

func notifyStatus(event string) {
	statusUpdates.notify(event)
}

func (statusEventListener) HandleEvent(event utils.Event) {
	notifyStatus(event.Name)
}

type ModuleStatus struct {
	Running bool                   `json:"running"`
	Info    map[string]interface{} `json:"info"`
}

func queryParams(r *http.Request) map[string]string {
	params := make(map[string]string)
	for k, v := range r.URL.Query() {
		if len(v) == 0 {
			continue
		}
		params[k] = v[0]
	}
	return params
}

func getStatus() map[string]interface{} {
	globalConfig, _ := core.GetGlobalConfig()
	status := make(map[string]interface{})
	status["global"] = map[string]interface{}{
		"config": globalConfig,
	}

	for name, module := range core.GetModules() {
		moduleStatus := make(map[string]interface{})
		moduleStatus["running"] = module.IsRunning()
		moduleStatus["config"], _ = module.GetConfig(nil)
		status[name] = moduleStatus
	}

	status["ipInfo"] = ipInfo.snapshot()

	return status
}

func statusHandler(w http.ResponseWriter, r *http.Request) {
	// Set SSE headers
	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("X-Accel-Buffering", "no")

	flusher, ok := w.(http.Flusher)
	if !ok {
		http.Error(w, "streaming unsupported", http.StatusInternalServerError)
		return
	}

	updates, unsubscribe := statusUpdates.subscribe()
	defer unsubscribe()

	data, _ := json.Marshal(getStatus())
	if _, err := fmt.Fprintf(w, "data: %s\n\n", data); err != nil {
		return
	}
	flusher.Flush()

	heartbeat := time.NewTicker(statusHeartbeatInterval)
	defer heartbeat.Stop()

	for {
		select {
		case event := <-updates:
			utils.LogLn("Received event:", event)
			data, _ := json.Marshal(getStatus())
			if _, err := fmt.Fprintf(w, "data: %s\n\n", data); err != nil {
				return
			}
			flusher.Flush()
		case <-heartbeat.C:
			if _, err := fmt.Fprint(w, ": keepalive\n\n"); err != nil {
				return
			}
			flusher.Flush()
		case <-r.Context().Done():
			return
		}
	}
}

func forceRefreshHandler(w http.ResponseWriter, _ *http.Request) {
	ipInfo.markEvent("force", time.Now())
	if !ipInfo.refresh() {
		notifyStatus("force")
		http.Error(w, "failed to refresh IP info", http.StatusBadGateway)
		return
	}
	notifyStatus("ip-info")
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(ipInfo.snapshot())
}

func versionHandler(w http.ResponseWriter, _ *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{"version": core.Version})
}

func getGlobalConfigHandler(w http.ResponseWriter, r *http.Request) {
	config, err := core.GetGlobalConfig()
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
	json.NewEncoder(w).Encode(config)
}

func saveGlobalConfigHandler(w http.ResponseWriter, r *http.Request) {
	var config map[string]interface{}
	err := json.NewDecoder(r.Body).Decode(&config)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	err = core.SaveGlobalConfig(config)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	w.WriteHeader(http.StatusOK)
}

// Get Module status
func getModuleStatusHandler(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	module := vars["module"]

	isRunning, err := core.GetModuleStatus(module)
	status := ModuleStatus{
		Running: isRunning,
		Info:    ipInfo.snapshot().Output,
	}
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
	json.NewEncoder(w).Encode(status)
}

// Enable Module
func enableModuleHandler(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	module := vars["module"]

	startNow := r.URL.Query().Get("start") == "true"

	err := core.EnableModule(module, startNow)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
	w.WriteHeader(http.StatusOK)
}

// Disable Module
func disableModuleHandler(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	module := vars["module"]

	stopNow := r.URL.Query().Get("stop") == "true"

	err := core.DisableModule(module, stopNow)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
	w.WriteHeader(http.StatusOK)
}

// Restart Module
func restartModuleHandler(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	module := vars["module"]

	err := core.RestartModule(module)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
	w.WriteHeader(http.StatusOK)
}

func getModuleConfigHandler(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	module := vars["module"]

	params := queryParams(r)
	config, err := core.GetModuleConfig(module, params)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
	json.NewEncoder(w).Encode(config)
}

// Save a config (new or existing)
func saveModuleConfigHandler(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	module := vars["module"]

	var params = queryParams(r)
	var config map[string]interface{}
	err := json.NewDecoder(r.Body).Decode(&config)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	err = core.SaveModuleConfig(module, params, config)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusOK)
}

type FileInfo struct {
	Name  string `json:"name"`
	Path  string `json:"path"`
	IsDir bool   `json:"isDir"`
}

// Helper function to sanitize paths
func sanitizePath(inputPath string) string {
	// Prevent navigating outside baseDir
	cleanPath := filepath.Clean("/" + inputPath) // Ensure the path starts with a slash
	return filepath.Join(core.VarDir, cleanPath)
}

func listFilesHandler(w http.ResponseWriter, r *http.Request) {
	relPath := r.URL.Query().Get("path")
	absPath := sanitizePath(relPath)

	files, err := os.ReadDir(absPath)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	var fileInfos []FileInfo
	for _, file := range files {
		if strings.HasSuffix(file.Name(), ".auth") {
			continue
		}
		fileInfos = append(fileInfos, FileInfo{
			Name:  file.Name(),
			Path:  filepath.Join(relPath, file.Name()), // Preserve the relative path for the client
			IsDir: file.IsDir(),
		})
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(fileInfos)
}

func fileContentHandler(w http.ResponseWriter, r *http.Request) {
	relPath := r.URL.Query().Get("path")
	absPath := sanitizePath(relPath)

	// Ensure the requested path is inside the base directory
	if !strings.HasPrefix(absPath, core.VarDir) || strings.HasSuffix(absPath, ".auth") {
		http.Error(w, "Access denied", http.StatusForbidden)
		return
	}

	content, err := os.ReadFile(absPath)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "text/plain")
	w.Write(content)
}

// Separate function to handle static files
func handleStaticFiles(r *mux.Router) {
	// Serve static files from /static and root (/)
	fs := http.FileServer(http.Dir(staticDir))
	noCache := func(handler http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			w.Header().Set("Cache-Control", "no-cache")
			handler.ServeHTTP(w, r)
		})
	}
	r.PathPrefix("/static/").Handler(noCache(http.StripPrefix("/static/", fs)))
	r.PathPrefix("/").Handler(noCache(http.StripPrefix("/", fs))) // Serve "/" from staticDir
}

func (i *ipInfoCache) HandleEvent(event utils.Event) {
	i.markEvent(event.Name, time.Now())
	notifyStatus(event.Name)

	if i.refresh() {
		notifyStatus("ip-info")
	}
}

func WebServer(port string) {
	utils.RegisterListener([]string{"vpn-up", "vpn-down"}, ipInfo)
	utils.RegisterListener([]string{"proxy-up", "proxy-down"}, statusEventListener{})

	ipInfo.markEvent("startup", time.Now())
	go func() {
		if ipInfo.refresh() {
			notifyStatus("ip-info")
		}
	}()

	// Create a new Gorilla Mux router
	r := mux.NewRouter()

	// Config-related routes
	r.HandleFunc("/api/status", statusHandler).Methods("GET")
	r.HandleFunc("/api/version", versionHandler).Methods("GET")
	r.HandleFunc("/api/force-refresh", forceRefreshHandler).Methods("GET")
	r.HandleFunc("/api/config", getGlobalConfigHandler).Methods("GET")
	r.HandleFunc("/api/config/save", saveGlobalConfigHandler).Methods("POST")

	// Handle file
	r.HandleFunc("/api/files", listFilesHandler)
	r.HandleFunc("/api/file", fileContentHandler)

	// Module
	r.HandleFunc("/api/{module}/status", getModuleStatusHandler).Methods("GET")
	r.HandleFunc("/api/{module}/enable", enableModuleHandler).Methods("POST")
	r.HandleFunc("/api/{module}/disable", disableModuleHandler).Methods("POST")
	r.HandleFunc("/api/{module}/restart", restartModuleHandler).Methods("POST")
	r.HandleFunc("/api/{module}/config", getModuleConfigHandler).Methods("GET")
	r.HandleFunc("/api/{module}/config/save", saveModuleConfigHandler).Methods("POST")

	// Custom module routes
	for _, module := range core.GetModules() {
		module.RegisterRoutes(r)
	}

	// Serve static files
	handleStaticFiles(r)

	// Start the server
	utils.LogF("Server starting on port %s\n", port)
	err := http.ListenAndServe(":"+port, r)
	if err != nil {
		utils.LogFatal(err)
	}
}

File: server/webserver/webserver.py
import os
import json
import threading
import time
from datetime import datetime
from flask import Flask, request, jsonify, Response, send_from_directory
import core
from core import get_modules, enable_module, disable_module, restart_module, get_module_config, save_module_config, get_global_config, save_global_config, Version
from utils import get_ip_info, log_ln, EventListener, register_listener

app = Flask(__name__, static_folder="../static", static_url_path="/")

class StatusNotifier:
    def __init__(self):
        self.subscribers = []
        self.mu = threading.Lock()
        
    def subscribe(self):
        q = []
        with self.mu:
            self.subscribers.append(q)

        def unsubscribe():
            with self.mu:
                if q in self.subscribers:
                    self.subscribers.remove(q)

        return q, unsubscribe
        
    def notify(self, event):
        with self.mu:
            for q in list(self.subscribers):
                q.append(event)

status_updates = StatusNotifier()
status_heartbeat_interval = 15

class IpInfoCache:
    def __init__(self):
        self.mu = threading.Lock()
        self.refresh_mu = threading.Lock()
        self.output = {}
        self.executed_at = None
        self.last_event = ""
        self.last_event_at = None
        
    def mark_event(self, name, at):
        with self.mu:
            self.last_event = name
            self.last_event_at = at
            
    def store(self, output, executed_at):
        with self.mu:
            self.output = output
            self.executed_at = executed_at
            
    def snapshot(self):
        with self.mu:
            stale = True
            if self.executed_at and self.last_event_at:
                stale = self.executed_at < self.last_event_at
            elif not self.executed_at:
                stale = True
            else:
                stale = False
                
            return {
                "output": dict(self.output),
                "executedAt": self.executed_at.isoformat() + "Z" if self.executed_at else None,
                "event": self.last_event,
                "eventAt": self.last_event_at.isoformat() + "Z" if self.last_event_at else datetime.now().isoformat() + "Z",
                "stale": stale
            }
            
    def refresh(self):
        with self.refresh_mu:
            updated = {}
            err = get_ip_info(updated)
            if err:
                return False
            self.store(updated, datetime.now())
            return True
            
    def handle_event(self, event):
        self.mark_event(event.name, datetime.now())
        notify_status(event.name)
        if self.refresh():
            notify_status("ip-info")

ip_info = IpInfoCache()

def notify_status(event):
    status_updates.notify(event)

class StatusEventListener(EventListener):
    def handle_event(self, event):
        notify_status(event.name)

def get_status():
    global_cfg, _ = get_global_config()
    status = {
        "global": {
            "config": global_cfg
        }
    }
    
    for name, module in get_modules().items():
        status[name] = {
            "running": module.is_running(),
            "config": module.get_config({})
        }
        
    status["ipInfo"] = ip_info.snapshot()
    return status

@app.route("/api/status")
def status_handler():
    def stream():
        q, unsubscribe = status_updates.subscribe()
        try:
            data = json.dumps(get_status())
            yield f"data: {data}\n\n"
            
            while True:
                if q:
                    event = q.pop(0)
                    log_ln(f"Received event: {event}")
                    data = json.dumps(get_status())
                    yield f"data: {data}\n\n"
                else:
                    yield ": keepalive\n\n"
                time.sleep(1) # Simple polling for SSE for KISS
        finally:
            unsubscribe()
            
    return Response(stream(), content_type="text/event-stream")

@app.route("/api/force-refresh")
def force_refresh_handler():
    ip_info.mark_event("force", datetime.now())
    if not ip_info.refresh():
        notify_status("force")
        return "failed to refresh IP info", 502
    notify_status("ip-info")
    return jsonify(ip_info.snapshot())

@app.route("/api/version")
def version_handler():
    return jsonify({"version": Version})

@app.route("/api/config")
def get_global_config_handler():
    cfg, err = get_global_config()
    if err:
        return str(err), 500
    return jsonify(cfg)

@app.route("/api/config/save", methods=["POST"])
def save_global_config_handler():
    config = request.json
    err = save_global_config(config)
    if err:
        return str(err), 500
    return "", 200

@app.route("/api/<module>/status")
def get_module_status_handler(module):
    mod = get_modules().get(module)
    if not mod:
        return "module not found", 404
    is_running = mod.is_running()
    return jsonify({
        "running": is_running,
        "info": ip_info.snapshot()["output"]
    })

@app.route("/api/<module>/enable", methods=["POST"])
def enable_module_handler(module):
    start_now = request.args.get("start") == "true"
    err = enable_module(module, start_now)
    if err:
        return str(err), 500
    return "", 200

@app.route("/api/<module>/disable", methods=["POST"])
def disable_module_handler(module):
    stop_now = request.args.get("stop") == "true"
    err = disable_module(module, stop_now)
    if err:
        return str(err), 500
    return "", 200

@app.route("/api/<module>/restart", methods=["POST"])
def restart_module_handler(module):
    err = restart_module(module)
    if err:
        return str(err), 500
    return "", 200

@app.route("/api/<module>/config")
def get_module_config_handler(module):
    params = request.args.to_dict()
    cfg, err = get_module_config(module, params)
    if err:
        return str(err), 500
    return jsonify(cfg)

@app.route("/api/<module>/config/save", methods=["POST"])
def save_module_config_handler(module):
    params = request.args.to_dict()
    config = request.json
    err = save_module_config(module, params, config)
    if err:
        return str(err), 500
    return "", 200

def sanitize_path(input_path):
    import posixpath
    clean_path = posixpath.normpath("/" + input_path).lstrip("/")
    return os.path.join(core.VarDir, clean_path)

@app.route("/api/files")
def list_files_handler():
    rel_path = request.args.get("path", "")
    abs_path = sanitize_path(rel_path)
    
    if not os.path.exists(abs_path):
        return str(Exception("not found")), 500
        
    file_infos = []
    try:
        for f in os.listdir(abs_path):
            if f.endswith(".auth"):
                continue
            f_path = os.path.join(abs_path, f)
            file_infos.append({
                "name": f,
                "path": os.path.join(rel_path, f).lstrip("/"),
                "isDir": os.path.isdir(f_path)
            })
        file_infos.sort(key=lambda x: x["name"])
    except Exception as e:
        return str(e), 500
        
    return jsonify(file_infos)

@app.route("/api/file")
def file_content_handler():
    rel_path = request.args.get("path", "")
    abs_path = sanitize_path(rel_path)
    
    if not abs_path.startswith(core.VarDir) or abs_path.endswith(".auth"):
        return "Access denied", 403
        
    try:
        with open(abs_path, "rb") as f:
            content = f.read()
        return Response(content, content_type="text/plain")
    except Exception as e:
        return str(e), 500

@app.route("/", defaults={'path': ''})
@app.route("/<path:path>")
def serve_static(path):
    if path != "" and os.path.exists(os.path.join(app.static_folder, path)):
        return send_from_directory(app.static_folder, path)
    else:
        return send_from_directory(app.static_folder, 'index.html')

def start_webserver(port):
    register_listener(["vpn-up", "vpn-down"], ip_info)
    register_listener(["proxy-up", "proxy-down"], StatusEventListener())
    register_listener(["apps-changed"], StatusEventListener())
    
    ip_info.mark_event("startup", datetime.now())
    threading.Thread(target=lambda: notify_status("ip-info") if ip_info.refresh() else None).start()
    
    for _, module in get_modules().items():
        module.register_routes(app)
        
    log_ln(f"Server starting on port {port}")
    # Run flask with threaded=True which is default, good for SSE
    app.run(host="0.0.0.0", port=int(port), debug=False, use_reloader=False)
